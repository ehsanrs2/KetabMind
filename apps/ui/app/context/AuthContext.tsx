'use client';

import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
  type ReactNode,
} from 'react';

type User = {
  id: string;
  name: string;
  email?: string;
  language?: string;
};

type AuthContextType = {
  user: User | null;
  loading: boolean;
  error: string | null;
  refresh: () => Promise<void>;
  logout: () => Promise<void>;
};

export const AuthContext = createContext<AuthContextType | undefined>(undefined);

async function fetchMe() {
  const response = await fetch('/api/me', {
    credentials: 'include',
  });

  if (!response.ok) {
    if (response.status === 401) {
      return null;
    }
    throw new Error('Failed to load user profile');
  }

  return (await response.json()) as User;
}

export function AuthProvider({ children }: { children: ReactNode }) {
  const [user, setUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);
  const mountedRef = useRef(true);

  useEffect(() => {
    mountedRef.current = true;
    return () => {
      mountedRef.current = false;
    };
  }, []);

  const applyResult = useCallback((nextUser: User | null, nextError: string | null) => {
    if (!mountedRef.current) {
      return;
    }
    setUser(nextUser);
    setError(nextError);
  }, []);

  const loadUser = useCallback(async () => {
    try {
      const data = await fetchMe();
      applyResult(data, null);
    } catch (err) {
      applyResult(null, err instanceof Error ? err.message : 'Unknown error');
    }
  }, [applyResult]);

  useEffect(() => {
    (async () => {
      setLoading(true);
      await loadUser();
      if (mountedRef.current) {
        setLoading(false);
      }
    })();
  }, [loadUser]);

  const refresh = useCallback(async () => {
    setLoading(true);
    await loadUser();
    if (mountedRef.current) {
      setLoading(false);
    }
  }, [loadUser]);

  const logout = useCallback(async () => {
    try {
      await fetch('/api/logout', {
        method: 'POST',
        credentials: 'include',
      });
    } catch (err) {
      console.warn('Logout failed', err);
    } finally {
      if (mountedRef.current) {
        setUser(null);
      }
    }
  }, []);

  const value = useMemo(
    () => ({
      user,
      loading,
      error,
      refresh,
      logout,
    }),
    [error, loading, logout, refresh, user],
  );

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export function useAuth() {
  const context = useContext(AuthContext);

  if (!context) {
    throw new Error('useAuth must be used within an AuthProvider');
  }

  return context;
}
