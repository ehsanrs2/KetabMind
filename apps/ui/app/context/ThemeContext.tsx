'use client';

import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

type Theme = 'light' | 'dark';

type ThemeContextValue = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
  toggleTheme: () => void;
};


const STORAGE_KEY = 'ketabmind.theme';

function getEnvDefaultTheme(): Theme {
  const raw =
    (process.env.NEXT_PUBLIC_THEME_DEFAULT ?? process.env.THEME_DEFAULT ?? 'light').toString();
  return normaliseTheme(raw) ?? 'light';
}

function normaliseTheme(value: unknown): Theme | null {
  if (typeof value !== 'string') {
    return null;
  }
  const lower = value.toLowerCase();
  return lower === 'dark' || lower === 'light' ? (lower as Theme) : null;
}

function readStoredTheme(): Theme | null {
  if (typeof window === 'undefined') {
    return null;
  }
  try {
    const stored = window.localStorage.getItem(STORAGE_KEY);
    return normaliseTheme(stored);
  } catch (error) {
    console.warn('Unable to read stored theme preference', error);
    return null;
  }
}

const ThemeContext = createContext<ThemeContextValue | undefined>(undefined);

export function ThemeProvider({
  children,
  initialTheme,
}: {
  children: React.ReactNode;
  initialTheme?: Theme;
}) {
  const defaultTheme = useMemo(() => initialTheme ?? getEnvDefaultTheme(), [initialTheme]);
  const [theme, setTheme] = useState<Theme>(() => {
    if (initialTheme) {
      return initialTheme;
    }
    return readStoredTheme() ?? defaultTheme;
  });

  useEffect(() => {
    if (initialTheme) {
      setTheme(initialTheme);
    }
  }, [initialTheme]);

  useEffect(() => {
    if (typeof document !== 'undefined') {
      document.body.dataset.theme = theme;
    }
    if (typeof window !== 'undefined') {
      try {
        window.localStorage.setItem(STORAGE_KEY, theme);
      } catch (error) {
        console.warn('Unable to persist theme preference', error);
      }
    }
  }, [theme]);

  useEffect(() => {
    if (typeof window === 'undefined' || initialTheme) {
      return;
    }
    const handleStorage = (event: StorageEvent) => {
      if (event.key !== STORAGE_KEY) {
        return;
      }
      const nextTheme = normaliseTheme(event.newValue);
      if (nextTheme) {
        setTheme((current) => (current === nextTheme ? current : nextTheme));
      }
    };
    const stored = readStoredTheme();
    if (stored) {
      setTheme((current) => (current === stored ? current : stored));
    }
    window.addEventListener('storage', handleStorage);
    return () => {
      window.removeEventListener('storage', handleStorage);
    };
  }, [initialTheme]);

  const toggleTheme = useCallback(() => {
    setTheme((current) => (current === 'light' ? 'dark' : 'light'));
  }, []);

  const value = useMemo<ThemeContextValue>(
    () => ({
      theme,
      setTheme,
      toggleTheme,
    }),
    [theme, toggleTheme],
  );

  return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>;
}

export function useTheme(): ThemeContextValue {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}

