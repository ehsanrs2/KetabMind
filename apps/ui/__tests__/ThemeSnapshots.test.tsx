import { render, screen } from '@testing-library/react';
import { ThemeProvider, useTheme } from '../app/context/ThemeContext';

function ThemeConsumer() {
  const { theme } = useTheme();
  return <div data-testid="current-theme">{theme}</div>;
}

describe('ThemeProvider snapshots', () => {
  beforeEach(() => {
    window.localStorage.clear();
    delete document.body.dataset.theme;
  });

  it('renders light theme snapshot', () => {
    const { asFragment } = render(
      <ThemeProvider initialTheme="light">
        <ThemeConsumer />
      </ThemeProvider>,
    );

    expect(screen.getByTestId('current-theme')).toHaveTextContent('light');
    expect(document.body.dataset.theme).toBe('light');
    expect(asFragment()).toMatchSnapshot('light-theme');
  });

  it('renders dark theme snapshot', () => {
    const { asFragment } = render(
      <ThemeProvider initialTheme="dark">
        <ThemeConsumer />
      </ThemeProvider>,
    );

    expect(screen.getByTestId('current-theme')).toHaveTextContent('dark');
    expect(document.body.dataset.theme).toBe('dark');
    expect(asFragment()).toMatchSnapshot('dark-theme');
  });
});

