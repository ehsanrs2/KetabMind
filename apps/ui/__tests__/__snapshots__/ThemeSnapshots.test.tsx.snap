// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider snapshots renders dark theme snapshot: dark-theme 1`] = `
<DocumentFragment>
  <div
    data-testid="current-theme"
  >
    dark
  </div>
</DocumentFragment>
`;

exports[`ThemeProvider snapshots renders light theme snapshot: light-theme 1`] = `
<DocumentFragment>
  <div
    data-testid="current-theme"
  >
    light
  </div>
</DocumentFragment>
`;
